### DISCUSSION TITLE
Welcome to tMAVEN Discussions!

### DISCUSSION THREAD

**Discussion by ckinzthompson** [GonzalezBiophysicsLab/tmaven#6] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by radny91** [comment]
Hi,

I've been trying to use tmaven and see how it works on my sm files. I've been using iSMS thus far so I exported my traces there to the SMD format. However, tmavern doesn't seem to read this format at all, any idea how can I make iSMS compatible with tmavern?

Best,
Rad

**Reply by ckinzthompson** [reply]
Hi Rad,
Glad to hear that you're trying out tMAVEN! We'll look into this and come up with a solution. The solution will probably be a permanent fix, so just FYI you might have to download an updated version of tMAVEN -- we'll see. I've opened an issue for this #7

**Reply by KorakRay** [reply]
Hi @radny91,

Welcome to the tMAVEN community! This has been an issue for other tMAVEN users migrating their traces from MATLAB-based softwares to tMAVEN. Typically, they report a reversal in the order of their data. As @ckinzthompson mentioned, we are looking into a more permanent fix for this issue.

In the meantime, a quick way to check for this error is to open the SMD file with the HDF5 viewer (under View>HDF5 viewer). A default Python-generated SMD should show up as
<img width="804" alt="image" src="https://github.com/GonzalezBiophysicsLab/tmaven/assets/8438932/83c98d7c-8475-40bd-9973-dd0d6a9af468">

in this case denoting that the L1-L9 dataset contains 590 trajectories (N) with 500 timepoints (T) and 2 colours (C). All tMAVEN data is stored in the NxTxC format, but we can change the order in which we load our data in the Preferences (View>Preferences, or Cmd/Ctrl+P). The specific preference io.axis_order controls the order in which each dimension is loaded.
<img width="278" alt="image" src="https://github.com/GonzalezBiophysicsLab/tmaven/assets/8438932/9a011678-7eb2-4b5c-b1ee-efcd07d10d14">


The default is [0,1,2], signifying that the first axis (0) is N, and so on. If your data is in the CxTxN format (due to C vs F ordering differences between MATLAB and Python), changing this preference to [2,1,0] will allow it to be loaded into tMAVEN without issue.

Let us know if this fixes your issue! Hopefully, this is a temporary work-around for now.
Regards,
Korak

**Reply by radny91** [reply]
Hi Korak,

Apologies for the late reply. Life in the lab has been busy. The axis order change didn't really work and there must be something inherently wrong with iSMS smd extensions in MATLAB. I bypassed this problem by exporting traces to txt files and writing a short script to save all of them as h5 file. Then, when loaded as a raw HDF5 dataset, tMAVEN reads all of the traces fine, giving the same output I'd observe in iSMS.

Radek

**Comment by dtanrio** [comment]
Hello! We have been trying to use tMAVEN in our lab to process our FRET traces. A common issue we have been running against is after using ebHMM to complete global HMM modeling of states, trying to run dwell time analysis based on these makes the program crash every time (when calculating). It accompanied by the following error: 

File "/Applications/anaconda3/lib/python3.8/site-packages/tmaven/controllers/modeler/dwells.py", line 22, in generate_dwells
    ind = int(np.argwhere(d[0] == means))
TypeError: only size-1 arrays can be converted to Python scalars
Abort trap: 6

Could you give us any tips on how we might resolve this issue/if this is a common problem?

Thank you!
Derin Tanrioven

**Reply by ckinzthompson** [reply]
Hi Derin, thanks for the details -- we'll take a look! @KorakRay, looks like dwells isn't properly handling eb results containers?

**Reply by ckinzthompson** [reply]
Hey Derin, we've updated tmaven to v 0.2.2 with this fix. Thanks for pointing it out! One thing to note is that empirical Bayes (EB) methods are "global" in a strange way. After estimating an appropriate prior that covers the entire ensemble, that prior is used to model each trace individually. In those resulting posteriors, there's no guarantee that the states are the same between the different traces -- this wouldn't be an issue anywhere other than dwell time analyses. That said, they are most likely the same, and 99% of the time this isn't an issue. I would just make sure to verify using a different method.

**Reply by dtanrio** [reply]
Hello,

Thank you so much for your quick reply! I still seem to be having issues plotting these dwells after calculating the dwell times on v0.2.2. It crashes again this time with the following error when the plot button is utilized (I believe we also ran into this error when we tried other methods to plot the dwells, so I'm not sure if this might have to do with our file format. We use SPARTAN traces.):

Traceback (most recent call last):
  File "/Applications/anaconda3/lib/python3.8/site-packages/tmaven/interface/ui_dwell.py", line 184, in <lambda>
    dwell_dialog.add_push_button('Plot', grid2, [1,1], fxn = lambda : dwell_dialog.plot_dwells())
  File "/Applications/anaconda3/lib/python3.8/site-packages/tmaven/interface/ui_dwell.py", line 140, in plot_dwells
    popplot_container(self.gui,self.gui.maven.plots.survival_dwell)
  File "/Applications/anaconda3/lib/python3.8/site-packages/tmaven/interface/ui_analysisplots.py", line 80, in __init__
    self.plot()
  File "/Applications/anaconda3/lib/python3.8/site-packages/tmaven/interface/ui_analysisplots.py", line 95, in plot
    self.maven_plot.plot(self.fig,self.ax)
  File "/Applications/anaconda3/lib/python3.8/site-packages/tmaven/controllers/analysis_plots/survival_dwell.py", line 90, in plot
    tau, surv = self.maven.modeler.get_survival_dwells(self.prefs['dwell_state'])
  File "/Applications/anaconda3/lib/python3.8/site-packages/tmaven/controllers/modeler/modeler.py", line 286, in get_survival_dwells
    tau, survival_norm = survival(np.array(dwells_model[str(state)]))
  File "/Applications/anaconda3/lib/python3.8/site-packages/numba/np/arraymath.py", line 605, in array_max_impl
    raise ValueError(MSG)
ValueError: zero-size array to reduction operation maximum which has no identity
Abort trap: 6


Thanks again!

**Reply by ckinzthompson** [reply]
I can't tell without seeing your data (feel free to email it to me colin.kinzthompson at rutgers.edu), but it looks like you have a state that is not populated and we did not protect against that situation (sorry!). How many states are you using in your model?
